src/core/queriesObserver.ts

```diff
@@ -64,78 +64,116 @@ export class QueriesObserver extends Subscribable<QueriesObserverListener> {
   }
 
   getOptimisticResult(queries: QueryObserverOptions[]): QueryObserverResult[] {
-    return queries.map((options, index) => {
-      const defaultedOptions = this.client.defaultQueryObserverOptions(options)
-      return this.getObserver(defaultedOptions, index).getOptimisticResult(
-        defaultedOptions
-      )
+    return this.getUpdatedObservers(queries).newResult
+  }
+
+  private getUpdatedObservers(
+    queries: QueryObserverOptions[],
+    notifyOptions?: NotifyOptions
+  ): QueryObserverUpdate {
+    const prevObservers = this.observers
+    const prevObserversMap = this.observersMap
+    const newResult: QueryObserverResult[] = []
+    const newObservers: QueryObserver[] = []
+    const newObserversMap: Record<string, QueryObserver> = {}
+
+    const defaultedQueryOptions = queries.map(options =>
+      this.client.defaultQueryObserverOptions(options)
+    )
+    const matchingObservers = defaultedQueryOptions
+      .map(options => {
+        const match = prevObserversMap[options.queryHash!]
+        if (match != null) {
+          match.setOptions(options, notifyOptions)
+          return match
+        }
+        return null
+      })
+      .filter(notNullOrUndefined)
+
+    const matchedQueryHashes = matchingObservers.map(
+      observer => observer?.options.queryHash
+    )
+    const unmatchedQueries = defaultedQueryOptions.filter(
+      options => !matchedQueryHashes.includes(options.queryHash)
+    )
+
+    const unmatchedObservers = prevObservers.filter(
+      prevObserver => !matchingObservers.includes(prevObserver)
+    )
+
+    const newlyMatchedOrCreatedObservers = unmatchedQueries.map(options => {
+      if (options.keepPreviousData && unmatchedObservers.length > 0) {
+        // use the first observer but no longer matched query to keep query data for any new queries
+        const firstObserver = unmatchedObservers.splice(0, 1)[0]
```

Comment:
splice mutates `unmatchedObservers`. It is a newly created array due to `prevObservers.filter`, but it would still be appreciated if we keep this immutable for readability. Or is this intended (I don't think it is since we abort the loop if we find the `firstObserver`..)

Reply:
You are right, it's better to keep this immutable. Simple indexing works here nicely as well, so changed the code to get rid of splice 👍